test/index.js

```diff
@@ -0,0 +1,274 @@
+  eq (parse (['--color']))
+     (Right (Pair ({color: true, email: Nothing, words: []})
+                  ([])));
+
+  eq (parse (['--colour']))
+     (Right (Pair ({color: true, email: Nothing, words: []})
+                  ([])));
+
+  eq (parse (['--email', 'mail@example.org']))
+     (Right (Pair ({color: false, email: Just ('mail@example.org'), words: []})
+                  ([])));
+
+  eq (parse (['--email', 'xxx@example.org', '--email', 'mail@example.org']))
+     (Right (Pair ({color: false, email: Just ('mail@example.org'), words: []})
+                  ([])));
+
+  eq (parse (['--word', 'foo']))
+     (Right (Pair ({color: false, email: Nothing, words: ['foo']})
+                  ([])));
+
+  eq (parse (['--word', 'foo', '--word', 'bar']))
+     (Right (Pair ({color: false, email: Nothing, words: ['foo', 'bar']})
+                  ([])));
+
+  eq (parse (['--word', 'foo', '--word', 'bar', '--word', 'baz']))
+     (Right (Pair ({color: false, email: Nothing, words: ['foo', 'bar', 'baz']})
+                  ([])));
+
+  eq (parse (['--word', 'foo', '--color', '--word', 'bar', '--email', 'mail@example.org', '--word', 'baz']))
+     (Right (Pair ({color: true, email: Just ('mail@example.org'), words: ['foo', 'bar', 'baz']})
+                  ([])));
+
+  eq (parse (['--word', '--']))
+     (Right (Pair ({color: false, email: Nothing, words: ['--']})
+                  ([])));
+
+  eq (parse (['--word', '--color']))
+     (Right (Pair ({color: false, email: Nothing, words: ['--color']})
+                  ([])));
+});
+
+test ('[success] short options (separate)', () => {
+  eq (parse (['-c', '-e', 'mail@example.org', '-w', 'foo', '-w', 'bar', '-w', 'baz']))
+     (Right (Pair ({color: true, email: Just ('mail@example.org'), words: ['foo', 'bar', 'baz']})
+                  ([])));
+
+  eq (parse (['-e', 'xxx@example.org', '--email', 'mail@example.org']))
+     (Right (Pair ({color: false, email: Just ('mail@example.org'), words: []})
+                  ([])));
+
+  eq (parse (['--email', 'xxx@example.org', '-e', 'mail@example.org']))
+     (Right (Pair ({color: false, email: Just ('mail@example.org'), words: []})
+                  ([])));
+
+  eq (parse (['-w', 'short', '--word', 'long', '-w', 'short', '--word', 'long']))
+     (Right (Pair ({color: false, email: Nothing, words: ['short', 'long', 'short', 'long']})
+                  ([])));
+
+  eq (parse (['-w', '--']))
+     (Right (Pair ({color: false, email: Nothing, words: ['--']})
+                  ([])));
+
+  eq (parse (['-w', '-c']))
+     (Right (Pair ({color: false, email: Nothing, words: ['-c']})
+                  ([])));
+});
+
+test ('[success] short options (combined)', () => {
+  eq (parse (['-ce', 'mail@example.org']))
+     (Right (Pair ({color: true, email: Just ('mail@example.org'), words: []})
+                  ([])));
+
+  eq (parse (['-wxc']))
+     (Right (Pair ({color: true, email: Nothing, words: ['-x']})
+                  ([])));
```

Comment:
This behaviour is surprising. What do you think *should* happen in this case, Aldwin?

Reply:
Oof, yeah.. that's very surprising. I'd have been more comfortable with that being an error like `-w requires a value`, possibly with additional context: `; instead saw -x`.

Reply:
Or maybe a more context aware error like: `you tried combining short option -w with a flag -x. When combining an option with flags, the option must be positioned last in the list of flags, and be followed by a value`.

Reply:
I guess most parsers would assign `xc` as the value to `-w`, but that's something you'd wish to avoid, right?

Reply:
That got me to think about your original problem statement here: https://github.com/sanctuary-js/sanctuary-argv/pull/1#discussion_r703528287

While it's true that `-wxc` *looks* ambiguous to the writer, the parser has additional information to reduce the ambiguity: namely whether a given argument is a flag or an option.

So the list of permutations is drastically reduced:

- `-x yz` only when `-x` is an option
- `-x -y z` only when `-x` is a flag and `-y` is an option
- `-x -y -z` only when `-x` is a flag and `-y` is a flag

Reply:
...wait, or maybe I missed the original point. ;)

Reply:
This now gives `Left ('-w requires a value')`, which makes sense. :)